### DISCUSSION TITLE
Rendering multiple scenes with different active cameras and resolutions

### DISCUSSION THREAD

**Discussion by beyond9thousand** [p2or/blender-loom#88] [Ideas]
Is there a way to do this even natively? If so, does loom have a way to achieve this?

**Comment by p2or** [comment]
Hi @beyond9thousand,

As of the current state, we can override the render settings including the 'camera' and 'resolution' property in the regular as well as the batch dialog, see the [presets](https://github.com/p2or/blender-loom?tab=readme-ov-file#render-presets) option. In order to render different scenes, you can either use the Blender's compositor or the [batch dialog](https://github.com/p2or/blender-loom?tab=readme-ov-file#batch-render-dialog-ctrlshiftaltf12) which provides a 'snapshot feature' that allows to create a (temporary) snapshot of the current blend file which automatically gets added to the queue:

<img width="762" alt="Screenshot 2024-03-12 at 09 40 59" src="https://github.com/p2or/blender-loom/assets/512368/8cb9e274-fbf8-4109-bcb1-d606db49c96a">
<br><br>

Should only take a few seconds to set up, depending on the project. Does this help and answer your question?

Cheers,
Christian

**Comment by beyond9thousand** [comment]
Blender's compositor only lets me decide the scene, but its specific properties like the resolution are not respected. The active scene always seems to take precedence

**Comment by p2or** [comment]
Hi @beyond9thousand,

as mentioned, the 'snapshot' operator should do in any case, just give it a whirl.

Cheers,
Christian

**Comment by beyond9thousand** [comment]
I have Scene1 and Scene2, how would I make this work exactly? I need to save the file with a Scene selected and then the add the snapshot I believe. But then how am I supposed to render the same file with the other Scene? Sorry but I am a little lost since the dialog box disappears if i click away to save the file with a different scene as currently active

Edit:
Was fiddling with the information available online and it seems that you can technically loop through all scenes in a blend file using python, thereby achieving the exact result i am referring to. could this functionality be added to loom as an option?

**Comment by p2or** [comment]
Hi @beyond9thousand,

> I have Scene1 and Scene2, how would I make this work exactly?

1. Select `Scene1` 
1. Open up the 'Batch dialog' and hit <kbd>Add Snapshot</kbd>
1. Select `Scene2`
1. Open up the 'Batch dialog' and hit <kbd>Add Snapshot</kbd>
...

All snapshots end up in a new folder called 'temp' (can be customized in the preferences). To render everything, either add a new scene to the existing `.blend` or create a dedicated `.blend` for rendering, open up the Batch dialog add the files within the temp folder to the list and render.

> seems that you can technically loop through all scenes in a blend file using python

Here you go: `for scene in D.scenes: print(scene.name)` but I don't see how that could help unless we loop through all the scenes and take snapshots ;) 

<img width="577" alt="Screenshot 2024-07-06 at 11 37 10" src="https://github.com/p2or/blender-loom/assets/512368/9e440b1f-2d17-4b56-ab5a-af7a675f481e">

```python
import bpy

context = bpy.context
current_scene = context.window.scene

for scene in bpy.data.scenes:
    ''' Assemble the filename '''
    filename = file_name="{}_{}".format(
            bpy.path.basename(bpy.data.filepath[:-6]),
            scene.name)
    ''' Switch the scene and call the operator '''
    context.window.scene = scene
    bpy.ops.loom.batch_snapshot(file_name=filename)
    
context.window.scene = current_scene
```




> could this functionality be added to loom as an option?

Sure, a more direct way to override the camera, layer or scene per render execution should be straight forward to implement but I have no idea how to add that properties within the 'Batch dialog' without destroying the current logic. If you have an idea or mockup for the UI, please let me know and let's discuss about why this is worth to implement in the first place.

Cheers,
Christian

**Comment by beyond9thousand** [comment]
The issue with the snapshot method as it turns out, is that the filepath for the output set in compositor gets messed up. This is because (personally) its more convenient and foolproof to set a  relative file path `//../Render/Path/`

This is especially true when I am working with a team or rendering parts of the same file from different machines as this happens to also be the most powerful way to use your addon in Blender.

Since the snapshot function creates another directory in the current directory (tmp), it creates an output in a lower level down the tree.

As for the functionality being added into the batch dialog, I think we can have a small button that can toggle on rendering for each scene (all Scenes in the given blend file get rendered, as opposed to the active scene only). Its default behavior would be to only render the active scene in the saved file.

![BlenderLoom_ScenesButton_Mockup](https://github.com/p2or/blender-loom/assets/77363063/656cb5a3-9d9e-4fd2-be46-352a2bbc058e)

**Reply by p2or** [reply]
Hi @beyond9thousand,

> The issue with the snapshot method as it turns out, is that the filepath for the output set in compositor gets messed up. This is because (personally) its more convenient and foolproof to set a relative file path `//../Render/Path/` ... Since the snapshot function creates another directory in the current directory (tmp), it creates an output in a lower level down the tree.

Thanks for pointing this out. As of [the latest commit](https://github.com/p2or/blender-loom/commit/efc2df2884343cd2f001aa6cfff353dd096fbf8a), all output filepaths gets converted into an absolute paths when taking a snapshot which should solve your issue. There is also a new operator that can be used to convert back to relative paths if required:

https://github.com/p2or/blender-loom/assets/512368/aadc2166-685f-4215-9524-ac8bda3c5238

Please download Loom directly from the [master branch](https://github.com/p2or/blender-loom) and report back.

> As for the functionality being added into the batch dialog, I think we can have a small button that can toggle on rendering for each scene (all Scenes in the given blend file get rendered, as opposed to the active scene only).

Why all scenes in one go? I personally think this is rarely needed, don't you? Otherwise, please elaborate. 

Cheers,
Christian

**Comment by beyond9thousand** [comment]
The reason why I required all the scenes to be rendered one by one is cases where two scenes have different cameras and resolutions. Specifically when working with forced perspective 3D content (which our Studio mainly works with)

I am attaching an example below:
![NutellaRegional_MultiScenes](https://github.com/p2or/blender-loom/assets/77363063/a81e823b-9485-402b-b3a1-215ebaaf8e56)

As you can see in this case, for the flat screens we would set up a camera perpendicular to the 3D frame/room but for something like a curved screen, we need to project the content onto a perspective unwrapped surface. The resolution for the curved screen would go through two runs.

At the same time, our content is connected across the 3 screens which means the most efficient way to work would be to have linked scenes.

This example is actually the exact content we were working on when I inquired you about the specific functionality

**Comment by p2or** [comment]
Hi @beyond9thousand,

got it, thanks. I think you will agree that rendering for 3 outdoor screens simultaneous is a bit special right? In general, it would be nice to make this add-on useful to everyone and in this regard I still see no advantage in adding a button to render all scenes one after the other. Beeing able to override the scene or the camera per item in the queue is an old request but Blender's UI is visually very limited when it comes to lists and looks crappy without a good idea which is my actual problem in this matter... Are you with me?

Nonetheless, I've already tried a lot e.g.:

https://github.com/p2or/blender-loom/assets/512368/a4092835-45e8-4fd5-855f-e21a21bb8f78

The best thing so far is to display the properties on demand and add an option to duplicate queue elements:

https://github.com/p2or/blender-loom/assets/512368/2439d0ef-7901-4039-bc7c-b5a4837affe9

If you'd like to help, you can download the [latest changes](https://github.com/p2or/blender-loom/tree/batch-overrides-ui), test it and report back. It works technically, but I'm still not completely satisfied with the interface though. If you have a better idea, let me know. 

Cheers,
Christian

**Comment by beyond9thousand** [comment]
This is splendid for the time being to be honest. Give me a few more days and I will let my team members give their own feedback on this as well.

**Comment by beyond9thousand** [comment]
On that note, any idea what this could be? One of the scenes in my file keeps triggering this issue with loom

```
Read blend: "C:\Users\beyond9thousand\Downloads\HSBC_UV_Unwrap_v2.blend"
init
Scene switch for render: 'UV Unwrap' in file: 'C:\Users\beyond9thousand\Downloads\HSBC_UV_Unwrap_v2.blend'
Traceback (most recent call last):
  File "C:\Users\beyond9thousand\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\loom.py", line 4033, in execute
    if scn.render.use_sequencer and len(scn.sequence_editor.sequences):
AttributeError: 'NoneType' object has no attribute 'sequences'
Error: Python: Traceback (most recent call last):
  File "C:\Users\beyond9thousand\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\loom.py", line 4033, in execute
    if scn.render.use_sequencer and len(scn.sequence_editor.sequences):
AttributeError: 'NoneType' object has no attribute 'sequences'
Location: C:\Users\beyond9thousand\scoop\apps\blender\current\4.0\scripts\modules\bpy\ops.py:109
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\beyond9thousand\scoop\apps\blender\current\4.0\scripts\modules\bpy\ops.py", line 109, in __call__
    ret = _op_call(self.idname_py(), kw)
RuntimeError: Error: Python: Traceback (most recent call last):
  File "C:\Users\beyond9thousand\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\loom.py", line 4033, in execute
    if scn.render.use_sequencer and len(scn.sequence_editor.sequences):
AttributeError: 'NoneType' object has no attribute 'sequences'
Location: C:\Users\beyond9thousand\scoop\apps\blender\current\4.0\scripts\modules\bpy\ops.py:109
```

[HSBC_UV_Unwrap_v2.zip](https://github.com/user-attachments/files/16199019/HSBC_UV_Unwrap_v2.zip)

**Comment by p2or** [comment]
Hi @beyond9thousand,

I guess that the original file was created in a very old version of Blender (< 2.65) or it has been saved incorrectly. You can either (1) disable the 'Sequencer' property under *Output > Post Processing* or (2) temporarily add a new strip of any type to the sequencer and then delete it immediately. If this happens frequently, we can add a test to Loom for such cases...  

Fixed file for Blender 3.6 LTS: [HSBC_UV_Unwrap_Blender3.6_v2.blend.zip](https://github.com/user-attachments/files/16201280/HSBC_UV_Unwrap_Blender3.6_v2.blend.zip) 

Cheers,
Christian

**Comment by beyond9thousand** [comment]
It was created in 4.0.2

I disabled sequencer properties and it worked, but not sure why I don't run into the same problem with the other scene (Perspective)

**Comment by beyond9thousand** [comment]
Do you think we can push it as an official release for the scene selector version or do you still have some reservations on it?

**Comment by p2or** [comment]
Hi @beyond9thousand,

> Do you think we can push it as an official release for the scene selector version

why this is a thing for you right now? 

> or do you still have some reservations on it?

In general, I'd like to work on a real project with it before merging the branch to be on the safe side.

Apart from that, I am still looking for a reliable way to exclude cameras that are not part of the selected scene and would like to add an option to the snapshot function to include all available scenes of the current file to the queue, but can live without it. I'm also still waiting for your promised feedback. How does it feel with the new features?

Cheers,
Christian

**Reply by beyond9thousand** [reply]
Ah my bad. I should have mentioned alongside that I like the modification you were able to make. It's a very flexible mode for the scenarios where it's easier to manage multiple scenes in the same file. 

Usually it would require me to save the file separately or use the temporary snapshot method but I like the scene switcher option much better.

As for excluding the cameras, I don't believe that's something you need to personally take care of via your plugin.

Different scenes maintain their own cameras and resolution etc respectively already. Blender's inbuilt stereoscopy option lets us render more than one camera on the same frame natively anyway. 

Please correct me if we're not on the same page regarding the camera topic you mentioned. 

My only concern is the sequencer in post processing issue which usually never gets triggered. Saying this because I've used your plugin on several projects so far and I've never gotten that error (except the last time I mentioned this) 

I've been running on Blender 4.0.2 since it's release in case you're curious.

**Reply by p2or** [reply]
Hi @beyond9thousand,

great! thanks for your feedback.

> As for excluding the cameras, I don't believe that's something you need to personally take care of via your plugin.

To avoid possible user errors, the camera selection should be dependent on the selection of the scene and at the moment all cameras of the blend file are listed in the dropdown which is a little drawback.

> My only concern is the sequencer in post processing issue which usually never gets triggered. Saying this because I've used your plugin on several projects so far and I've never gotten that error (except the last time I mentioned this)

I haven't changed anything in this regard and this has nothing to do with _Loom_. From a technical point of view, the entire 'Sequencer' and its properties does not exist in your file for whatever reason... Based on this knowledge, I can only say that your file is either corrupted or was originally created in a very old version of Blender. Either way, now you know how to fix it right? To avoid such errors as much as possible, I would recommend using the LTS versions in a production environment.

Cheers,
Christian

**Reply by beyond9thousand** [reply]
Need to test the latest LTS before adopting it into our production. And yes you are right, for now I can just bypass the sequencer issue altogether if I ever come across it since I know what needs to be done.

**Reply by beyond9thousand** [reply]
Its been a minute

Happy to share that the feature you added on the fly based on our discussion has proven to be extremely useful in the long run. The ability to batch-assign the same Blender file across multiple instances while selecting different scenes for each has been a significant time saver, especially when distributing workloads to higher-performance machines during crunch periods!

**Reply by p2or** [reply]
Hi @beyond9thousand,

great! If you haven't noticed, I [updated the feature branch](https://github.com/p2or/blender-loom/tree/batch-overrides-ui) six weeks ago to speed up the camera and scene selection within the dialog itself. Also, the latest changes from the master branch are now included.

However, I haven't had the time to test it extensively. If you have the time, please download the [latest version](https://github.com/p2or/blender-loom/archive/refs/heads/batch-overrides-ui.zip) and let me know your thoughts so we can hopefully merge it in the upcoming weeks.

Cheers,
Christian
